cmd/attest.go

```diff
@@ -0,0 +1,193 @@
+)
+
+var attestationKeys = map[string]map[tpm2.Algorithm]func(rw io.ReadWriter) (*client.Key, error){
+	"AK": {
+		tpm2.AlgRSA: client.AttestationKeyRSA,
+		tpm2.AlgECC: client.AttestationKeyECC,
+	},
+	"gceAK": {
+		tpm2.AlgRSA: client.GceAttestationKeyRSA,
+		tpm2.AlgECC: client.GceAttestationKeyECC,
+	},
+}
+
+var attestCmd = &cobra.Command{
+	Use:   "attest",
+	Short: "Create a remote attestation report",
+	Long: `Gather information for remote attestation.
+
+The Attestation report contains a quote on all available PCR banks ,a way to validate 
+the quote, and a TCG Event Log (Linux only).
+
+Use --key to specify the type of attestation key. It can be gceAK for GCE attestation
+key or AK for a custom attestation key. By default it uses gceAK.
+--algo flag override the public key algorithm for attestation key. If not provided then
+by default rsa is used.
+--nonce attaches even length extra data to the report and guarantees a fresh quote.
+--teenonce attaches a 64 bytes extra data to the attestation report of hardware and 
+guarantees a fresh quote.
+--format flag specify the type of output file. It can be "binarypb" or "textproto".
+By default it uses binarypb.
+--output flag provides file name to store attestation report in the format specified
+by --format flag.
+`,
+	Args: cobra.NoArgs,
+	RunE: func(cmd *cobra.Command, args []string) error {
+
+		rwc, err := openTpm()
+		if err != nil {
+			return err
+		}
+		defer rwc.Close()
+
+		createFunc := attestationKeys[key][keyAlgo]
+
+		attestationKey, err := createFunc(rwc)
+		if err != nil {
+			return fmt.Errorf("failed to create attestation key: %v", err)
+		}
+		defer attestationKey.Close()
+
+		attestopts := client.AttestOpts{}
+		attestopts.Nonce = nonce
+		if len(teenonce) != 0 {
+			attestopts.TEENonce = teenonce
+			attestopts.TEEDevice, err = client.CreateSevSnpDevice()
+			if err != nil {
+				return fmt.Errorf("collecting TEE attestation report: %v", err)
+			}
+		}
+
+		if ExternalTPM != nil {
+			if elg, ok := ExternalTPM.(client.EventLogGetter); ok {
+				attestopts.TCGEventLog, err = elg.EventLog()
+				if err != nil {
+					return fmt.Errorf("failed to retrieve TCG Event Log: %w", err)
+				}
+			}
+		}
+
+		attestation, err := attestationKey.Attest(attestopts)
+		if err != nil {
+			return fmt.Errorf("failed to collect attestation report: %v", err)
+		}
+
+		if key == "gceAK" {
+			getMetadata()
+			addGceInstanceInfo(attestation)
+		}
+
+		if len(output) != 0 {
+			var out []byte
+			if format == "binarypb" {
+				out, err = proto.Marshal(attestation)
+				if err != nil {
+					return fmt.Errorf("failed to marshal attestation proto: %v", attestation)
+				}
+			} else {
+				out = []byte(prototext.Format(attestation))
+			}
+			if _, err := dataOutput().Write(out); err != nil {
+				return fmt.Errorf("failed to write attestation proto: %v", err)
+			}
+		}
+		return nil
+	},
+}
+
+func getMetadata() error {
+
+	var err error
+	gceProjectID, err = metadata.ProjectID()
```

Comment:
Please add tests against a fake MDS for every item we pull, and double check we get the correct information.

Reply:
As of now , we are unable to make a GCE key using ExternalTPM.
And without GCE key , we can't retrieve the GCEInstanceInfo as per the current implementation.